Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm

```diff
@@ -148,73 +178,78 @@
         m_webPage->removeTextAnimationForAnimationID(animationID);
 }
 
-#if PLATFORM(MAC)
-static WebCore::CharacterRange newlyReplacedCharacterRange(WebCore::CharacterRange superRange, WebCore::CharacterRange previousRange)
+static WebCore::CharacterRange remainingCharacterRange(WebCore::CharacterRange totalRange, WebCore::CharacterRange previousRange)
 {
     auto location = previousRange.location + previousRange.length;
-    auto length = superRange.length - previousRange.length;
-    if (superRange.length < previousRange.length) {
+    auto length = totalRange.length - previousRange.length;
+    if (totalRange.length < previousRange.length) {
         ASSERT_NOT_REACHED();
         return WebCore::CharacterRange { 0, 0 };
     }
     return WebCore::CharacterRange { location, length };
 };
-#endif
 
 void TextAnimationController::addInitialTextAnimation(const WTF::UUID& sessionUUID)
 {
     auto initialAnimationUUID = WTF::UUID::createVersion4();
-    auto sessionRange = contextRangeForSessionWithID(sessionUUID);
+    auto animatingRange = unreplacedRangeForSessionWithID(sessionUUID);
 
-    if (!sessionRange) {
-        ASSERT_NOT_REACHED();
+    if (!animatingRange || animatingRange->collapsed()) {
         return;
     }
+    auto textIndicatorData = createTextIndicatorForRange(*animatingRange);
+    if (!textIndicatorData)
+        return;
 
-    auto textIndicatorData = createTextIndicatorForRange(*sessionRange);
     m_webPage->addTextAnimationForAnimationID(initialAnimationUUID, { TextAnimationType::Initial, WTF::UUID(WTF::UUID::emptyValue) }, *textIndicatorData);
 
-    m_initialAnimations.add(sessionUUID, initialAnimationUUID);
+    m_initialAnimations.set(sessionUUID, initialAnimationUUID);
 }
 
 
-void TextAnimationController::addSourceTextAnimation(const WTF::UUID& sessionUUID, const WebCore::CharacterRange& currentReplacedRange)
+void TextAnimationController::addSourceTextAnimation(const WTF::UUID& sessionUUID, const WebCore::CharacterRange& replacingRange, const String string, WTF::CompletionHandler<void(void)>&& completionHandler)
 {
 #if PLATFORM(MAC)
-    auto currentlyStyledRange = m_alreadyReplacedRanges.getOptional(sessionUUID);
-    if (!currentlyStyledRange)
-        return;
+    auto previouslyReplacedRange = m_alreadyReplacedRanges.getOptional(sessionUUID);
+    auto replaceCharacterRange = replacingRange;
+
+    if (previouslyReplacedRange) {
+        if ((*previouslyReplacedRange)->range == replacingRange && (*previouslyReplacedRange)->string == string)
```

Comment:
```suggestion
        if (previouslyReplacedRange->range == replacingRange && previouslyReplacedRange->string == string)
```

Reply:
May even be nice to pass in a `ReplacedRangeAndString` as the parameter, so you can directly compare to that, but I won't hold this PR up for that.

Reply:
So, the reason this is (*previouslyReplacedRange)->range and not previouslyReplacedRange->range is because this is a double-wrapped optional. yes, I know it's crazy, I would like to fix it, but this is where we are right now.